### PR TITLE
feat(taipei-h3): onboard Taipei Hash House Harriers (台北捷兔)

### PR DESCRIPTION
**Pull request by johnrclem**
## Summary
Onboards **Taipei Hash House Harriers (台北捷兔)** — Taiwan's oldest hash, founded **13 March 1973**, run **#2779** as of June 2026. Second Taipei-area kennel after the already-live Taiwan H3 (`twh3-tw`). Full onboard: new `TaipeiHashAdapter` + kennel/alias/source seed + self-hosted logo. No region work (Taiwan COUNTRY + Taipei METRO already on `main`).

- **Type:** full onboard · **Adapter:** new `TaipeiHashAdapter` (static Cheerio over the server-rendered `run_site.php` PHP hareline)
- **kennelCode:** `taipei-h3` · **Source:** `https://www.taipeihash.com.tw/run_site.php` (HTML_SCRAPER, trust 6, daily, `upcomingOnly:true`)

## Why a bespoke adapter (not config-only)
Three source-specific quirks: bilingual ZH/EN cells, **run-number-anchored year inference**, and **PII phone stripping** from hare cells.

### Year-inference design 🔴
The page shows year-less `MM/DD` dates across a ~6-month window that includes ~23 **past** runs (#2756 = 01/03 is ~5 months before today). A naive today-anchored bidirectional rollover would push those Jan/Feb history rows a full year into the **future**. Instead:
1. Anchor on the row nearest *today* (a strictly-weekly kennel ⇒ this is the current run).
2. Expect every other row at `anchorDate + (runNumber − anchorRun) × 7 days`.
3. Resolve each row's year to the candidate (`y−1 / y / y+1`) whose `MM/DD` lands closest to that expectation, validating each via a `Date.UTC` round-trip.

### Fail-loud guard
A clean fetch that yields **0 events** pushes an `errors[]` + `errorDetails.parse` entry, so a markup drift can't "succeed" with `events: []` and let `reconcile` cancel live runs — a brand-new source has a 0 baseline the zero-event health alert won't catch.

### Other adapter notes
- Parses only `<table.events-table>` rows; the mobile `div.mobile-event-card` duplicates use a different DOM shape and are naturally skipped, with a dedup-by-run-number Map as belt-and-suspenders.
- PII phone strip via precise `<span class="phone">` removal + a bounded ReDoS-safe regex fallback.
- `title` left undefined → `merge.ts` synthesizes `"Taipei H3 Trail #N"`. `startTime` constant `"15:00"`. Maps shortlinks stored verbatim as `locationUrl` (no coords; merge geocodes the bilingual place text). Reuses `filterEventsByWindow` + `normalizeHaresField` from `adapters/utils`.

## Live verification ✅ (per `.claude/rules/live-verification.md`)
Ran `adapter.fetch()` against the production site (fixture also captured from the real DOM):
- **27 events**, **0 errors**, date range **2026-01-03 → 2026-07-04**
- Year resolution correct across deep history: **#2756 → 2026-01-03** (not 2027), **#2762 (02/14 生日) → 2026-02-14**, **#2779 → 2026-06-13**, **#2782 → 2026-07-04**
- **0 unmatched** kennelTags · **0 phone leaks** in hares · **27/27** Maps shortlinks · `startTime` 15:00 on every row · 4 upcoming

## Checks
- `npx tsc --noEmit` clean for changed files (2 unrelated pre-existing `@vercel/blob` env errors only)
- `npm run lint` — 0 errors
- `npm test` — full suite green (9061 passed); new `taipei-hash.test.ts` covers parse, dedup, year inference incl. a Dec→Jan boundary case, PII strip, and the fail-loud guard
- Reviewed via `/simplify` (reuse + refactor applied) and `/code-review` (3 theoretical edge cases surfaced; all declined as non-issues given verified-clean live data)

## Deep-dive checklist
- [x] logo (self-hosted `public/kennel-logos/taipei-h3.jpg`; confirmed JPEG via Content-Type + magic bytes)
- [x] foundedYear (1973) · [x] socials (FB group; IG/Twitter none — sibling Taiwan H3's are deliberately not reused)
- [x] schedule (Saturday 15:00 weekly; Sunday directory discrepancy noted in scheduleNotes — kennel's own page is authoritative)
- [x] hashCash (NT$200, beer included) · [x] description · [x] source live-verified
- [x] history depth assessed (single page ~6 months; Drive-PDF archive out of scope, no backfill script)
- [x] coord sanity (shortlinks only → geocode; no default-pin trap) · [x] end times (none)
- [x] kennelCode collision-checked (`taipei-h3` clear; bare "TH3" alias omitted — globally taken)
- [x] source-kennel guard set · [x] region already seeded (no `region.ts` edits)

## Post-merge (separate runbook)
`git pull` on main → verify files landed → `npx prisma db seed` (additive, no new regions) → trigger a scrape → spot-check `/kennels/taipei-h3` for ~26 runs (#2756→#2782), current #2779.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by johnrclem** [comment]
@claude @codex review

File: src/adapters/html-scraper/taipei-hash.ts [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+const START_TIME = "15:00"; // "每星期六下午 3:00 起跑"
+const WEEK_MS = 7 * 24 * 60 * 60 * 1000;
+const NOON_HOUR = 12;
+
+const DATE_RE = /(\d{1,2})\/(\d{1,2})/; // MM/DD — ignores any trailing event tag
+const RUN_RE = /(\d+)/;
+// Strip a trailing phone number if the dedicated <span class="phone"> was absent
+// (markup-drift fallback). Single char class, bounded — ReDoS-safe.
+const PHONE_RE = /0\d[\d\s-]{6,15}/g;
+const MAPS_HREF = "a[href*='maps.app.goo.gl']";
```
[the remark this answers is not in the document]
Comment (johnrclem) [reply src/adapters/html-scraper/taipei-hash.ts]: Applied in a128eaa5 — broadened `MAPS_HREF` to also match `goo.gl/maps` and `google.com/maps`. The site uses `maps.app.goo.gl` 27/27 today; this is cheap forward-resilience. Extraction was already href-based.

File: src/adapters/html-scraper/taipei-hash.ts [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+// Two source-specific quirks drive a bespoke adapter (vs config-only):
+//   1. Year-less MM/DD dates over a ~6-month window that includes deep history.
+//      A naive today-anchored rollover mis-dates the Jan/Feb history rows a year
+//      into the future — instead we anchor on run number (strictly weekly) and
+//      resolve each row's year to whatever lands nearest its expected date.
+//   2. Hare cells carry a phone number (PII) in a dedicated <span class="phone">.
+
+const KENNEL_TAG = "taipei-h3";
+const DEFAULT_URL = "https://www.taipeihash.com.tw/run_site.php";
+const START_TIME = "15:00"; // "每星期六下午 3:00 起跑"
```
[the remark this answers is not in the document]
Comment (johnrclem) [reply src/adapters/html-scraper/taipei-hash.ts:30]: Declining — this looks like a mix-up with the sibling Taiwan H3 (`twh3-tw`), whose hares-off is 14:30. Taipei H3's own live hareline says **每星期六下午 3:00 起跑** (Saturday 3:00 PM start); the only 2:00 on the page is the suggested early-arrival/warm-up time (下午 2:00 左右到達集合地點). There is no 2:30/14:30 anywhere in run_site.php. 15:00 is correct.

File: src/adapters/html-scraper/taipei-hash.ts [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+  clone.find("br").replaceWith(" ");
+  const text = clone.text().replace(PHONE_RE, "").replace(/\s+/g, " ").trim();
+  return normalizeHaresField(text);
+}
+
+/** Google Maps shortlink from the 記號起點 cell, stored verbatim (no coords). */
+function parseMapsUrl(marksCell: Cheerio<Element>): string | undefined {
+  const href = marksCell.find(MAPS_HREF).first().attr("href");
+  return href && href.trim().length > 0 ? href.trim() : undefined;
+}
```
[the remark this answers is not in the document]
Comment (johnrclem) [reply src/adapters/html-scraper/taipei-hash.ts:112]: Applied in 8cca95a4 — `parseMapsUrl` now parses the href with `new URL()`, requires `https:`, and checks the hostname against an allowlist (maps.app.goo.gl / goo.gl / google.com / www.google.com / maps.google.com). Added a unit test asserting a foreign host with an embedded Maps substring is rejected. Verified live: 27/27 real links still pass, 0 non-https.

File: src/adapters/html-scraper/taipei-hash.ts [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+): ParsedRow[] {
+  const rows: ParsedRow[] = [];
+  $("table.events-table tbody tr").each((i, tr) => {
+    try {
+      const tds = $(tr).children("td");
+      if (tds.length < 5) return; // header / spacer row
+      const runNumber = parseRunNumber(tds.eq(0));
+      const md = parseMonthDay(tds.eq(1));
+      if (runNumber === null || md === null) return;
+      rows.push({
```
[the remark this answers is not in the document]
Comment (johnrclem) [reply src/adapters/html-scraper/taipei-hash.ts:171]: Applied in 8cca95a4 — a 5-cell row that yields no run number/date now pushes an `errors[]` + `errorDetails.parse` diagnostic (with rawText) instead of returning silently. Header/spacer rows are still skipped quietly by the earlier `tds.length < 5` guard, so this only fires on genuine markup drift.

File: src/adapters/html-scraper/taipei-hash.ts [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+export class TaipeiHashAdapter implements SourceAdapter {
+  type = "HTML_SCRAPER" as const;
+
+  async fetch(source: Source, options?: { days?: number }): Promise<ScrapeResult> {
+    const url = source.url || DEFAULT_URL;
+    const refDate = new Date();
+
+    const page = await fetchHTMLPage(url);
```
[the remark this answers is not in the document]
Comment (johnrclem) [reply src/adapters/html-scraper/taipei-hash.ts:254]: Good catch, but declining on this PR — it's shared infra: the missing default timeout is in `safeFetch`'s direct-fetch branch and affects every HTML-scraper adapter via `fetchHTMLPage`, not just Taipei. Fixing it here would be an out-of-scope cross-cutting change. Tracked as a separate follow-up to add `signal: init.signal ?? AbortSignal.timeout(45_000)` to the direct branch (mirroring the proxy branch).

[the remark this answers is not in the document]
Comment (johnrclem) [reply src/adapters/html-scraper/taipei-hash.ts:254]: @coderabbitai yes please — open a tracking issue for adding `signal: init.signal ?? AbortSignal.timeout(45_000)` to the direct-fetch branch of `safeFetch`. Thanks!
